File: frontend/src/types/filter-schema.ts
/**
 * Filter schema types for dynamic filter configuration.
 * 
 * These types mirror the backend FilterSchema models and enable
 * type-safe filter configuration throughout the application.
 */

export type FilterFieldType = 'string' | 'boolean' | 'entity' | 'column'

export type OptionsSource = 'entities' | 'columns'

/**
 * Metadata for a single filter configuration field.
 */
export interface FilterFieldMetadata {
  name: string
  type: FilterFieldType
  required: boolean
  default?: any
  description: string
  placeholder: string
  options_source?: OptionsSource
}

/**
 * Complete schema for a filter type.
 */
export interface FilterSchema {
  key: string
  display_name: string
  description: string
  fields: FilterFieldMetadata[]
}

/**
 * Runtime filter configuration.
 * Type is required, all other fields are dynamic based on the filter schema.
 */
export interface FilterConfig {
  type: string
  [key: string]: any
}

File: backend/app/models/filter_schema.py
"""Filter schema models for API responses."""

from typing import Any, Literal

from pydantic import BaseModel, Field


class FilterFieldMetadataResponse(BaseModel):
    """Filter field metadata for API response."""

    name: str = Field(..., description="Field name")
    type: Literal["string", "boolean", "entity", "column"] = Field(..., description="Field type")
    required: bool = Field(False, description="Whether field is required")
    default: Any = Field(None, description="Default value")
    description: str = Field("", description="Field description")
    placeholder: str = Field("", description="Placeholder text")
    options_source: Literal["entities", "columns"] | None = Field(None, description="Source for autocomplete options")


class FilterSchemaResponse(BaseModel):
    """Filter schema for API response."""

    key: str = Field(..., description="Filter identifier")
    display_name: str = Field(..., description="Human-readable name")
    description: str = Field(..., description="Filter description")
    fields: list[FilterFieldMetadataResponse] = Field(..., description="Configuration fields")
